File: Programmers/C++/모음 사전.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>
#define FOR(i,n) for(int i=0;i<(n);++i)
#define FORS(i,s,n) for(int i=(s);i<(n);++i)

using namespace std;

int res;
char ch[5];
vector<string> v;

void dfs(int cnt,vector<int>& cur) {
    if(cur.size()==cnt) {
        string s = "";
        for(int i: cur) s+=ch[i];
        v.push_back(s);
        return;
    }
    FOR(i,5) {
        cur.push_back(i);
        dfs(cnt,cur);
        cur.pop_back();
    }
}

int solution(string word) {
    ch[0] = 'A';
    ch[1] = 'E';
    ch[2] = 'I';
    ch[3] = 'O';
    ch[4] = 'U';

    FORS(cnt,1,6) {
        vector<int> tmp;
        dfs(cnt,tmp);
    }
    sort(v.begin(),v.end());
    FOR(i,v.size()) if(v[i]==word) {
        res=i+1;
        break;
    }
    return res;
}

File: Programmers/Swift/모음 사전.swift
import Foundation

let ch = ["A","E","I","O","U"]
var res = 0
var v = [String]()

func dfs(_ cnt:Int, _ cur: inout [Int]) {
    if cnt == cur.count {
        var str = ""
        for x in cur {
            str += ch[x]
        }
        v.append(str)
        return
    }
    for i in 0..<5 {
        cur.append(i)
        dfs(cnt, &cur)
        cur.removeLast()
    }
}

func solution(_ word:String) -> Int {
    
    for cnt in 1...5 {
        var tmp = [Int]()
        dfs(cnt, &tmp)
    }
    v.sort()
    
    for i in 0..<v.count {
        if v[i] == word {
            res = i+1
            break
        }
    }
    
    return res
}
